server/src/main/java/org/opensearch/search/SearchService.java

```diff
@@ -1621,22 +1619,34 @@ private CanMatchResponse canMatch(ShardSearchRequest request, boolean checkRefre
                     request.getClusterAlias()
                 );
                 Rewriteable.rewrite(request.getRewriteable(), context, false);
+
+                if (hasRefreshPending) {
+                    final FieldSortBuilder sortBuilder = FieldSortBuilder.getPrimaryFieldSortOrNull(request.source());
+                    final MinAndMax<?> minMax = sortBuilder != null ? FieldSortBuilder.getMinMaxOrNull(context, sortBuilder) : null;
+                    return new CanMatchResponse(true, minMax);
+                }
+
                 final boolean aliasFilterCanMatch = request.getAliasFilter().getQueryBuilder() instanceof MatchNoneQueryBuilder == false;
-                FieldSortBuilder sortBuilder = FieldSortBuilder.getPrimaryFieldSortOrNull(request.source());
-                MinAndMax<?> minMax = sortBuilder != null ? FieldSortBuilder.getMinMaxOrNull(context, sortBuilder) : null;
-                boolean canMatch;
-                if (canRewriteToMatchNone(request.source())) {
-                    QueryBuilder queryBuilder = request.source().query();
-                    canMatch = aliasFilterCanMatch && queryBuilder instanceof MatchNoneQueryBuilder == false;
-                } else {
-                    // null query means match_all
-                    canMatch = aliasFilterCanMatch;
+                if (aliasFilterCanMatch == false) {
+                    return new CanMatchResponse(false, null);
+                }
+
+                // null query means match_all
+                boolean canMatch = canRewriteToMatchNone(request.source()) == false
+                    || request.source().query() instanceof MatchNoneQueryBuilder == false;
+                if (canMatch == false) {
+                    return new CanMatchResponse(false, null);
```

Comment:
Can we assume that `CanMatchResponse(false, null)` equivalent to `CanMatchResponse(false,<non-null>)`?

Reply:
I think yes, `CanMatchPreFilterSearchPhase` extract the min/max values of each shard response and use them to pre-sort the shards prior to running the subsequent phases. However, the unmatched shard will be skipped in the following phases, so its min/max value will have no impact on the query task.